### PR TITLE
CTP-3137 Add button to remove mapping in dashboard page

### PR DESCRIPTION
**Pull request by aydevworks**
Deployed on 44-clc.

### CODE REVIEW

**Comment by stuartlamour** [comment]
Hi @aydevworks this looks like a lot of js code for a simple button with a confirm/cancel?

Other option include:

make the button a form, on submit use:
onclick="return confirm('Are you sure you want to remove FOO?')"

or follow the same pattern as the transfer marks button, using a single modal and passing data?

**Comment by aydevworks** [comment]
> Hi @aydevworks this looks like a LOT of js code for a simple button with confirm/cancel options?
> 
> Other option include:
> 
> make the button a form, on submit use:
> 
> onclick="return confirm('Are you sure you want to remove FOO?')" or if you need it to look pretty follow the same pattern as the transfer marks button, using a single modal in the template and passing data?

Thanks, updated.

**Comment by aydevworks** [comment]
Re-deployed on 44-clc.

**Comment by stuartlamour** [comment]
Hi, @aydevworks,

For MVP is it possible to just output this where you currently output the remove button:

 <form action="/page.php?param=foo" method="post" class="d-inline">
   <input type="hidden" name="foo" value="foo">
   
    <button type="submit" class="btn btn-primary" onclick="return confirm('Are you sure you want to remove FOO?')">
      Remove <span class="sr-only">foo name</span>
    </button>
 </form>

Then when form is posted with any data, php deals with it, without any extra intermediate interception and handling via javascript?

I'd be tempted to just do that to start.

If you do need a bootstrap modal instead of the standard ui confirm that is more complex, but possible (using js to trigger the form submit from the modal submit button).

**Comment by aydevworks** [comment]
> Hi, @aydevworks,
> 
> For MVP is it possible to just output this where you currently output the remove button:
> 
> ```
> <form action="/page.php?param=foo" method="post" class="d-inline">
>    <input type="hidden" name="foo" value="foo">
>    
>     <button type="submit" class="btn btn-primary" onclick="return confirm('Are you sure you want to remove FOO?')">
>       Remove <span class="sr-only">foo name</span>
>     </button>
>  </form>
> ```
> 
> Then when form is posted with any data, php deals with it, without any extra intermediate interception and handling via javascript?
> 
> I'd be tempted to just do that to start.
> 
> If you do need a bootstrap modal instead of the standard ui confirm that is more complex, but possible (using js to trigger the form submit from the modal submit button).

Hi @stuartlamour, I thought following the pattern of the transfer marks button was one of your suggestions in your previous comment?

Does it have any issue with the current approach like performance or something else?

Otherwise, I would prefer to keep the current code. It put the "remove source" function under the external folder together with other external functions, it seems odd to me to create another php file in somewhere else to do the "remove source" function.

File: classes/external/remove_mapping.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ */
+class remove_mapping extends external_api {
+    /**
+     * Returns description of method parameters.
+     *
+     * @return external_function_parameters
+     */
+    public static function execute_parameters() {
+        return new external_function_parameters([
+            'courseid' => new external_value(PARAM_INT, 'Coruse ID', VALUE_REQUIRED),
```
[the remark this answers is not in the document]
Comment (aydevworks) [reply classes/external/remove_mapping.php]: Thanks, updated.

**Comment by stuartlamour** [comment]
> Hi @stuartlamour, I thought following the pattern of the transfer marks button was one of your suggestions in your previous comment?
> 
> Does it have any issue with the current approach like performance or something else?
> 
> Otherwise, I would prefer to keep the current code. It put the "remove source" function under the external folder together with other external functions, it seems odd to me to create another php file in somewhere else to do the "remove source" function.

Hi @aydevworks,

Appols for confusion - by following the marks transfer pattern, i meant for how the modal was generated/populated :(

I'm not great at seeing how the SITS php files relate, but i can see a php function remove_mapping(int $courseid, int $mappingid)?

in dashboard.php could we not just add something like (terrible psudocode, sorry!):

`$removething = optional_param('removething', null, PARAM_INT);

$worldsaved = false; // Flag for save success message.
// Remove thing.
if ($removething) {
    $mappingid = required_param('mappingid', null, PARAM_INT);
    $saved = path::remove_mapping(COURSE->id, $mappingid);
    if ($saved) {
($worldsaved = true;
// Carry on.
}
}``

I may be being dumb here, as i've not followed the code all the way down, apologies.

Generally it is not good practice to use js to do things the web can do natively. It can create accessibility and performance issues.

If this is a pain, just go with the js approach and we can look at any implications later.

**Comment by aydevworks** [comment]
Replaced by this no JS version: https://github.com/ucl-isd/moodle-local_sitsgradepush/pull/76

**Pull request closed without merging**